internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala

```diff
@@ -51,78 +52,115 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile with
 
     def apply(unit: CompilationUnit): Unit = {
       if (!unit.isJava) {
-        // build dependencies structure
-        val sourceFile = unit.source.file.file
+        // Process dependencies if name hashing is enabled, fail otherwise
         if (global.callback.nameHashing) {
-          val dependencyExtractor = new ExtractDependenciesTraverser
-          dependencyExtractor.traverse(unit.body)
-
-          dependencyExtractor.memberRefDependencies foreach processDependency(context = DependencyByMemberRef)
-          dependencyExtractor.inheritanceDependencies foreach processDependency(context = DependencyByInheritance)
-          dependencyExtractor.localInheritanceDependencies foreach processDependency(context = LocalDependencyByInheritance)
-          processTopLevelImportDependencies(dependencyExtractor.topLevelImportDependencies)
+          val dependencyProcessor = new DependencyProcessor(unit)
+          val dependencyTraverser = new DependencyTraverser(dependencyProcessor)
+          // Traverse symbols in compilation unit and register all dependencies
+          dependencyTraverser.traverse(unit.body)
         } else {
           throw new UnsupportedOperationException(Feedback.NameHashingDisabled)
         }
-        /*
-         * Registers top level import dependencies as coming from a first top level class/trait/object declared
-         * in the compilation unit.
-         * If there's no top level template (class/trait/object def) declared in the compilation unit but `deps`
-         * is non-empty, a warning is issued.
-         */
-        def processTopLevelImportDependencies(deps: Iterator[Symbol]): Unit = if (deps.nonEmpty) {
-          val classOrModuleDef = firstClassOrModuleDef(unit.body)
-          classOrModuleDef match {
-            case Some(classOrModuleDef) =>
-              val sym = classOrModuleDef.symbol
-              val firstClassSymbol = if (sym.isModule) sym.moduleClass else sym
-              deps foreach { dep =>
-                processDependency(context = DependencyByMemberRef)(ClassDependency(firstClassSymbol, dep))
-              }
-            case None => reporter.warning(unit.position(0), Feedback.OrphanTopLevelImports)
-          }
+      }
+    }
+  }
+
+  private class DependencyProcessor(unit: CompilationUnit) {
+    private def firstClassOrModuleDef(tree: Tree): Option[Tree] = {
```

Comment:
Please, note that this code has been moved from the end of the file to the place where it's used. I decided not to change this implementation because there's not a `collectFirst` or `find` primitive in the reflect tree API and this is probably the most efficient one. I don't think it's worth it to reimplement it.